File: Trees/10. Count Good Nodes in Binary Tree/solution.scala
/** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left:
  * TreeNode = null, _right: TreeNode = null) { var value: Int = _value var
  * left: TreeNode = _left var right: TreeNode = _right }
  */
object Solution {
  def goodNodes(root: TreeNode): Int = {
    var result: Int = 0
    def dfs(node: TreeNode, currMax: Int): Unit = {
      if (node == null) return
      var newMax = currMax
      if (node.value >= currMax) {
        result += 1
        newMax = node.value
      }
      dfs(node.left, newMax)
      dfs(node.right, newMax)
    }
    dfs(root, 0)
    result
  }
}

File: Trees/10. Count Good Nodes in Binary Tree/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn good_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Solution::dfs(&root, None)
    }
    fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, max: Option<i32>) -> i32 {
        if let Some(n) = node {
            let val = n.borrow().val;
            let m = Some(max.map_or(val, |m| m.max(val)));
            (if m == Some(val) { 1 } else { 0 })
                + Solution::dfs(&n.borrow().left, m)
                + Solution::dfs(&n.borrow().right, m)
        } else {
            0
        }
    }
}
